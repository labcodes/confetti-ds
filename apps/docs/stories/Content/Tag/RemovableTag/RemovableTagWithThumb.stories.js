import React from "react";

import { TAG_COLORS } from "../../../../../../packages/confetti-ds/src/constants";
import { Button, RemovableTag as Component } from "../../../../../../packages/confetti-ds/src";

export default {
  title: "Content/Tag/Removable Tag",
  component: Component,
  argTypes: {
    color: {
      control: { type: "select", options: TAG_COLORS },
    },
    skin: {
      control: { type: "inline-radio" },
    },
    icon: {
      control: false,
    },
    thumbSrc: {
      control: false,
    },
  },
};

export const RemovableTagWithThumb = (args) => {
  const [showTag, setShowTag] = React.useState(true);

  return (
    <React.Fragment>
      { showTag ?
        <Component {...args} onClick={() => setShowTag(false)} />
        :
        <div>
          <Button text="Restore tag" onClick={() => setShowTag(true)} size="small"/>
        </div> }
    </React.Fragment>
  )
}
RemovableTagWithThumb.args = {
  text: "demo tag",
  thumbSrc: "http://avatars3.githubusercontent.com/u/1887591",
};
