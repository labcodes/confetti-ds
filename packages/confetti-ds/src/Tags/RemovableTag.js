import React from "react";
import PropTypes from "prop-types";
import { isEmpty } from "lodash";

import AbstractTag from "./AbstractTag";
import Icon from "../Icon";
import { ICON_TYPES, TAG_COLORS } from "../constants";

export default class RemovableTag extends React.Component {
  static propTypes = {
    /** This is the Tag's text. */
    text: PropTypes.string.isRequired,
    /** Source of the thumbnail to be rendered. Won't render a thumbnail if not passed to the component. */
    thumbSrc: PropTypes.string,
    /** Type of the icon to be rendered. Won't render an icon if not passed to the component. */
    icon: PropTypes.oneOf(ICON_TYPES),
    /** Sets Tag's color. */
    color: PropTypes.oneOf(TAG_COLORS),
    /** Skin of the the rendered Tag. */
    skin: PropTypes.oneOf(["pale", "vivid"]),
    /** Sets an outline style. */
    isOutline: PropTypes.bool,
    /** Disables the Tag. Will be read by screen readers. When true, will override `disabled`. */
    ariaDisabled: PropTypes.bool,
    /** Disables the Tag. Won't be read by screen readers. */
    disabled: PropTypes.bool,
    /** Callback to be executed when the Tag is clicked. */
    onClick: PropTypes.func,
  };

  static defaultProps = {
    thumbSrc: "",
    icon: undefined,
    color: undefined,
    skin: "pale",
    isOutline: false,
    disabled: false,
    ariaDisabled: false,
    onClick: () => {},
  };

  constructor(props) {
    super(props);
    this.checkThumbAndIcon();
  }

  componentDidUpdate() {
    this.checkThumbAndIcon();
  }

  removeIcon = () => (
    <span className="lab-tag__remove-icon-wrapper">
      <Icon
        type="remove"
        color="black-75"
        size="petit"
        className="lab-tag--remove-icon"
      />
    </span>
  );

  thumb = () => {
    const { thumbSrc } = this.props;
    return thumbSrc ? (
      <img className="lab-tag__thumb" src={thumbSrc} alt="" />
    ) : undefined;
  };

  icon = () => {
    const { icon } = this.props;
    return icon ? (
      <Icon
        type={icon}
        color="black-75"
        size="petit"
        className="lab-tag--left-icon"
      />
    ) : undefined;
  };

  checkThumbAndIcon() {
    const errorMessage =
      "`RemovableTag` can't be initialized with both `thumb` and `icon` props.";
    const { thumbSrc, icon } = this.props;
    if (!isEmpty(thumbSrc) && !isEmpty(icon)) {
      throw Error(errorMessage);
    }
  }

  render() {
    const {
      text,
      thumbSrc,
      icon,
      color,
      skin,
      isOutline,
      disabled,
      ariaDisabled,
      onClick,
    } = this.props;
    return (
      <AbstractTag
        className={`lab-tag--removable${`${
          icon ? ` lab-tag--has-left-icon` : ""
        }${thumbSrc ? ` lab-tag--has-thumb` : ""}`}`}
        text={text}
        thumbSrc={thumbSrc}
        icon={icon}
        color={color}
        skin={skin}
        isOutline={isOutline}
        disabled={!ariaDisabled && disabled}
        ariaDisabled={ariaDisabled}
        onClick={onClick}
        renderPrefix={this.thumb() || this.icon()}
        renderSuffix={this.removeIcon()}
      />
    );
  }
}
