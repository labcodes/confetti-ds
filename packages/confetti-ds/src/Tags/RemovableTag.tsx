import React, { SyntheticEvent, useMemo } from "react";
import { isEmpty } from "lodash";

import AbstractTag from "./AbstractTag";
import Icon from "../Icon";
import { IconTypes, TagTypes } from "../constants";

type BaseRemovableTagProps = {
  /** This is the Tag's text. */
  text: string;
  /** Type of the icon to be rendered. Won't render an icon if not passed to the component. Can't have both 'icon' and 'thumbSrc' at the same time. */
  icon?: IconTypes;
  /** Sets Tag's color. */
  color?: TagTypes;
  /** Skin of the rendered Tag. */
  skin?: "pale" | "vivid";
  /** Sets an outline style. */
  isOutline?: boolean;
  /** Disables the Tag. Will be read by screen readers. When true, will override `disabled`. */
  ariaDisabled?: boolean;
  /** Disables the Tag. Won't be read by screen readers. */
  disabled?: boolean;
  /** Callback to be executed when the Tag is clicked. */
  onClick?: (event?: SyntheticEvent) => any;
};

type RemovableTagProps =
  | (BaseRemovableTagProps & {
      /** Source of the thumbnail to be rendered. Won't render a thumbnail if not passed to the component. Can't have both 'icon' and 'thumbSrc' at the same time. */
      thumbSrc?: undefined;
      /** Source of the thumbnail to be rendered. Won't render a thumbnail if not passed to the component. Can't have both 'icon' and 'thumbSrc' at the same time. */
      thumbAlt?: undefined;
    })
  | (BaseRemovableTagProps & {
      /** Source of the thumbnail to be rendered. Won't render a thumbnail if not passed to the component. Can't have both 'icon' and 'thumbSrc' at the same time. */
      thumbSrc: string;
      /** Source of the thumbnail to be rendered. Won't render a thumbnail if not passed to the component. Can't have both 'icon' and 'thumbSrc' at the same time. */
      thumbAlt: string;
    });

export default function RemovableTag({
  text,
  icon,
  color,
  skin = "pale",
  isOutline = false,
  ariaDisabled = false,
  disabled = false,
  onClick = () => {},
  thumbSrc = "",
  thumbAlt = "",
}: RemovableTagProps) {
  const removeIcon = () => (
    <span className="lab-tag__remove-icon-wrapper">
      <Icon
        type="remove"
        color="black-75"
        size="petit"
        className="lab-tag--remove-icon"
      />
    </span>
  );

  const thumb = () =>
    thumbSrc.length ? (
      <img className="lab-tag__thumb" src={thumbSrc} alt={thumbAlt} />
    ) : undefined;

  const handleIcon = () =>
    icon ? (
      <Icon
        type={icon}
        color="black-75"
        size="petit"
        className="lab-tag--left-icon"
      />
    ) : undefined;

  const checkThumbAndIcon = () => {
    const errorMessage =
      "`RemovableTag` can't be initialized with both `thumb` and `icon` props.";
    if (!isEmpty(thumbSrc) && !isEmpty(icon)) {
      throw Error(errorMessage);
    }
  };

  useMemo(() => {
    checkThumbAndIcon();
  }, [icon, thumbSrc]);

  return (
    <AbstractTag
      className={`lab-tag--removable${`${
        icon ? ` lab-tag--has-left-icon` : ""
      }${thumbSrc ? ` lab-tag--has-thumb` : ""}`}`}
      text={text}
      color={color}
      skin={skin}
      isOutline={isOutline}
      disabled={!ariaDisabled && disabled}
      ariaDisabled={ariaDisabled}
      onClick={onClick}
      renderPrefix={thumb() || handleIcon()}
      renderSuffix={removeIcon()}
      tabIndex={0}
    />
  );
}
