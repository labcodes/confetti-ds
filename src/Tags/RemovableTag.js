import React, { useMemo } from "react";
import PropTypes from "prop-types";
import { isEmpty } from "lodash";

import AbstractTag from "./AbstractTag";
import Icon from "../Icon";
import { ICON_TYPES, TAG_COLORS } from "../constants";

export default function RemovableTag({
  text,
  thumbSrc,
  icon,
  color,
  skin,
  isOutline,
  disabled,
  ariaDisabled,
  onClick,
}) {
  const removeIcon = () => (
    <span className="lab-tag__remove-icon-wrapper">
      <Icon
        type="remove"
        color="black-75"
        size="petit"
        className="lab-tag--remove-icon"
      />
    </span>
  );

  const thumb = () =>
    thumbSrc ? (
      <img className="lab-tag__thumb" src={thumbSrc} alt="" />
    ) : undefined;

  const handleIcon = () =>
    icon ? (
      <Icon
        type={icon}
        color="black-75"
        size="petit"
        className="lab-tag--left-icon"
      />
    ) : undefined;

  const checkThumbAndIcon = () => {
    const errorMessage =
      "`RemovableTag` can't be initialized with both `thumb` and `icon` props.";
    if (!isEmpty(thumbSrc) && !isEmpty(icon)) {
      throw Error(errorMessage);
    }
  };

  useMemo(() => {
    checkThumbAndIcon();
  }, [icon, thumbSrc]);

  return (
    <AbstractTag
      className={`lab-tag--removable${`${
        icon ? ` lab-tag--has-left-icon` : ""
      }${thumbSrc ? ` lab-tag--has-thumb` : ""}`}`}
      text={text}
      thumbSrc={thumbSrc}
      icon={icon}
      color={color}
      skin={skin}
      isOutline={isOutline}
      disabled={!ariaDisabled && disabled}
      ariaDisabled={ariaDisabled}
      onClick={onClick}
      renderPrefix={thumb() || handleIcon()}
      renderSuffix={removeIcon()}
      tabIndex="0"
    />
  );
}

RemovableTag.propTypes = {
  /** This is the Tag's text. */
  text: PropTypes.string.isRequired,
  /** Source of the thumbnail to be rendered. Won't render a thumbnail if not passed to the component. Can't have both 'icon' and 'thumbSrc' at the same time. */
  thumbSrc: PropTypes.string,
  /** Type of the icon to be rendered. Won't render an icon if not passed to the component. Can't have both 'icon' and 'thumbSrc' at the same time. */
  icon: PropTypes.oneOf(ICON_TYPES),
  /** Sets Tag's color. */
  color: PropTypes.oneOf(TAG_COLORS),
  /** Skin of the rendered Tag. */
  skin: PropTypes.oneOf(["pale", "vivid"]),
  /** Sets an outline style. */
  isOutline: PropTypes.bool,
  /** Disables the Tag. Will be read by screen readers. When true, will override `disabled`. */
  ariaDisabled: PropTypes.bool,
  /** Disables the Tag. Won't be read by screen readers. */
  disabled: PropTypes.bool,
  /** Callback to be executed when the Tag is clicked. */
  onClick: PropTypes.func,
};

RemovableTag.defaultProps = {
  thumbSrc: "",
  icon: undefined,
  color: undefined,
  skin: "pale",
  isOutline: false,
  disabled: false,
  ariaDisabled: false,
  onClick: () => {},
};
