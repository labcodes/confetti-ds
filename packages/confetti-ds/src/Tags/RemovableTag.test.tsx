import React from "react";
import renderer from "react-test-renderer";
import { mount, shallow } from "enzyme";

import RemovableTag from "./RemovableTag";

describe("RemovableTag", () => {
  it("renders with base props", async () => {
    const renderedComponent = renderer
      .create(<RemovableTag text="Test render RemovableTag" />)
      .toJSON();
    expect(renderedComponent).toMatchSnapshot();

    const wrapper = shallow(<RemovableTag text="Test RemovableTag" />).html();
    expect(wrapper).toContain("lab-tag--removable");
  });

  it("renders as disabled", async () => {
    const renderedComponent = renderer
      .create(
        <RemovableTag text="Test render disabled RemovableTag" disabled />
      )
      .toJSON();
    expect(renderedComponent).toMatchSnapshot();

    const wrapper = shallow(
      <RemovableTag text="Test disabled RemovableTag" disabled />
    ).html();
    expect(wrapper).toContain("lab-tag--disabled");
  });

  it("renders as outline", async () => {
    const renderedComponent = renderer
      .create(
        <RemovableTag text="Test render outline RemovableTag" isOutline />
      )
      .toJSON();
    expect(renderedComponent).toMatchSnapshot();

    const wrapper = shallow(
      <RemovableTag text="Test outline RemovableTag" isOutline />
    ).html();
    expect(wrapper).toContain("lab-tag--outline");
  });

  it("renders with a green color", async () => {
    const renderedComponent = renderer
      .create(
        <RemovableTag text="Test render green RemovableTag" color="green" />
      )
      .toJSON();
    expect(renderedComponent).toMatchSnapshot();

    const wrapper = shallow(
      <RemovableTag text="Test green RemovableTag" color="green" />
    ).html();
    expect(wrapper).toContain("lab-tag--green");
  });

  it("calls prop.onClick when clicked", async () => {
    const mockOnClick = jest.fn();
    const shallowRemovableTag = shallow(
      <RemovableTag text="Test click on RemovableTag" onClick={mockOnClick} />
    );
    expect(mockOnClick.mock.calls.length).toEqual(0);
    shallowRemovableTag.simulate("click");
    expect(mockOnClick.mock.calls.length).toEqual(1);
  });

  it("doesn't trigger onClick if ariaDisabled", async () => {
    const mockOnClick = jest.fn();
    const mountedComponent = mount(
      <RemovableTag ariaDisabled text="Test" onClick={mockOnClick} />
    );

    expect(mockOnClick).not.toBeCalled();
    mountedComponent.find("span").at(0).simulate("click");
    expect(mockOnClick).not.toBeCalled();
  });

  it("does not render if passing both `thumb` and `icon`", async () => {
    expect(() => {
      shallow(
        <RemovableTag
          text="Test to not render RemovableTag with thumb and icon"
          icon="calendar"
          thumbSrc="fake-thumb"
          thumbAlt="fake-thumb-alt"
        />
      );
    }).toThrow(
      "`RemovableTag` can't be initialized with both `thumb` and `icon` props."
    );
  });
});
