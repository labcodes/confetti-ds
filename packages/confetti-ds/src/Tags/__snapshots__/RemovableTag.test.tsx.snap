// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RemovableTag renders as disabled 1`] = `
<span
  className="lab-tag lab-tag--removable lab-tag--disabled lab-tag--pale"
  onClick={[Function]}
  onKeyDown={[Function]}
  role="button"
  tabIndex={0}
>
  Test render disabled RemovableTag
  <span
    className="lab-tag__remove-icon-wrapper"
  >
    <span
      className="lab-icon lab-icon--Dismiss  lab-tag--remove-icon"
    >
      <svg
        aria-hidden={true}
        className="___12fm75w_v8ls9a0 f1w7gpdv fez10in fg4l7m0"
        fill="currentColor"
        height="20"
        viewBox="0 0 20 20"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="m4.09 4.22.06-.07a.5.5 0 0 1 .63-.06l.07.06L10 9.29l5.15-5.14a.5.5 0 0 1 .63-.06l.07.06c.18.17.2.44.06.63l-.06.07L10.71 10l5.14 5.15c.18.17.2.44.06.63l-.06.07a.5.5 0 0 1-.63.06l-.07-.06L10 10.71l-5.15 5.14a.5.5 0 0 1-.63.06l-.07-.06a.5.5 0 0 1-.06-.63l.06-.07L9.29 10 4.15 4.85a.5.5 0 0 1-.06-.63l.06-.07-.06.07Z"
          fill="currentColor"
        />
      </svg>
    </span>
  </span>
</span>
`;

exports[`RemovableTag renders as outline 1`] = `
<span
  className="lab-tag lab-tag--removable lab-tag--outline lab-tag--pale"
  onClick={[Function]}
  onKeyDown={[Function]}
  role="button"
  tabIndex={0}
>
  Test render outline RemovableTag
  <span
    className="lab-tag__remove-icon-wrapper"
  >
    <span
      className="lab-icon lab-icon--Dismiss  lab-tag--remove-icon"
    >
      <svg
        aria-hidden={true}
        className="___12fm75w_v8ls9a0 f1w7gpdv fez10in fg4l7m0"
        fill="currentColor"
        height="20"
        viewBox="0 0 20 20"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="m4.09 4.22.06-.07a.5.5 0 0 1 .63-.06l.07.06L10 9.29l5.15-5.14a.5.5 0 0 1 .63-.06l.07.06c.18.17.2.44.06.63l-.06.07L10.71 10l5.14 5.15c.18.17.2.44.06.63l-.06.07a.5.5 0 0 1-.63.06l-.07-.06L10 10.71l-5.15 5.14a.5.5 0 0 1-.63.06l-.07-.06a.5.5 0 0 1-.06-.63l.06-.07L9.29 10 4.15 4.85a.5.5 0 0 1-.06-.63l.06-.07-.06.07Z"
          fill="currentColor"
        />
      </svg>
    </span>
  </span>
</span>
`;

exports[`RemovableTag renders with a green color 1`] = `
<span
  className="lab-tag lab-tag--removable lab-tag--green-pale"
  onClick={[Function]}
  onKeyDown={[Function]}
  role="button"
  tabIndex={0}
>
  Test render green RemovableTag
  <span
    className="lab-tag__remove-icon-wrapper"
  >
    <span
      className="lab-icon lab-icon--Dismiss  lab-tag--remove-icon"
    >
      <svg
        aria-hidden={true}
        className="___12fm75w_v8ls9a0 f1w7gpdv fez10in fg4l7m0"
        fill="currentColor"
        height="20"
        viewBox="0 0 20 20"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="m4.09 4.22.06-.07a.5.5 0 0 1 .63-.06l.07.06L10 9.29l5.15-5.14a.5.5 0 0 1 .63-.06l.07.06c.18.17.2.44.06.63l-.06.07L10.71 10l5.14 5.15c.18.17.2.44.06.63l-.06.07a.5.5 0 0 1-.63.06l-.07-.06L10 10.71l-5.15 5.14a.5.5 0 0 1-.63.06l-.07-.06a.5.5 0 0 1-.06-.63l.06-.07L9.29 10 4.15 4.85a.5.5 0 0 1-.06-.63l.06-.07-.06.07Z"
          fill="currentColor"
        />
      </svg>
    </span>
  </span>
</span>
`;

exports[`RemovableTag renders with base props 1`] = `
<span
  className="lab-tag lab-tag--removable lab-tag--pale"
  onClick={[Function]}
  onKeyDown={[Function]}
  role="button"
  tabIndex={0}
>
  Test render RemovableTag
  <span
    className="lab-tag__remove-icon-wrapper"
  >
    <span
      className="lab-icon lab-icon--Dismiss  lab-tag--remove-icon"
    >
      <svg
        aria-hidden={true}
        className="___12fm75w_v8ls9a0 f1w7gpdv fez10in fg4l7m0"
        fill="currentColor"
        height="20"
        viewBox="0 0 20 20"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="m4.09 4.22.06-.07a.5.5 0 0 1 .63-.06l.07.06L10 9.29l5.15-5.14a.5.5 0 0 1 .63-.06l.07.06c.18.17.2.44.06.63l-.06.07L10.71 10l5.14 5.15c.18.17.2.44.06.63l-.06.07a.5.5 0 0 1-.63.06l-.07-.06L10 10.71l-5.15 5.14a.5.5 0 0 1-.63.06l-.07-.06a.5.5 0 0 1-.06-.63l.06-.07L9.29 10 4.15 4.85a.5.5 0 0 1-.06-.63l.06-.07-.06.07Z"
          fill="currentColor"
        />
      </svg>
    </span>
  </span>
</span>
`;
